# Fake News Detection

In [2]:
# Import Necessary Libraries
import numpy as np
import pandas as pd
import itertools
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import PassiveAggressiveClassifier
from sklearn.metrics import accuracy_score, confusion_matrix

In [3]:
# Read the data
df = pd.read_csv('news.csv')

# Get the shape and the head of the data
df.shape
df.head()

Unnamed: 0                                              title  \
0        8476                       You Can Smell Hillary’s Fear   
1       10294  Watch The Exact Moment Paul Ryan Committed Pol...   
2        3608        Kerry to go to Paris in gesture of sympathy   
3       10142  Bernie supporters on Twitter erupt in anger ag...   
4         875   The Battle of New York: Why This Primary Matters   

                                                text label  
0  Daniel Greenfield, a Shillman Journalism Fello...  FAKE  
1  Google Pinterest Digg Linkedin Reddit Stumbleu...  FAKE  
2  U.S. Secretary of State John F. Kerry said Mon...  REAL  
3  — Kaydee King (@KaydeeKing) November 9, 2016 T...  FAKE  
4  It's primary day in New York and front-runners...  REAL

In [4]:
# DataFlair - Get the labels from the DataFrame
labels = df.label
labels.head()

0    FAKE
1    FAKE
2    REAL
3    FAKE
4    REAL
Name: label, dtype: object

In [5]:
# DataFlair - Split the dataset
# Split the dataset into training and testing sets
x_train, x_test, y_train, y_test = train_test_split(df['text'], labels, test_size=0.2, random_state=7)

In [6]:
# DataFlair - Initialize a TfidfVectorizer
# This helps fit and transform the vectorizer on the train set, and transform the vectorized on the test set.
tfidf_vectorizer = TfidfVectorizer(stop_words='english', max_df=0.7)

# DataFlair - Fit and transform train set, transform test set
tfidf_train = tfidf_vectorizer.fit_transform(x_train)
tfidf_test = tfidf_vectorizer.transform(x_test)

### Initialize a PassiveAggressiveClassifier. Helping to fit this  on tfidf_train and y_train.
### We'll predict on the test set frmo the TfidfVectorizer and calculate the accuracy with accuracy_score() from sklearn.metrics.

In [7]:
# DataFlair - Initialize a PassiveAggressiveClassifier
pac = PassiveAggressiveClassifier(max_iter=50)
pac.fit(tfidf_train, y_train)

# DataFlair - Predict on the test set and calculate accuracy
y_pred = pac.predict(tfidf_test)
score = accuracy_score(y_test, y_pred)
print(f'Accuracy: {round(score*100,2)}%')

Accuracy: 92.98%


- We got an accuracy of 92.98%. 
- We can print out a confusion matrix to gain insight into the number of true and false positives and negatives.

In [8]:
# DataFlair - Build confusion matrix
confusion_matrix(y_test, y_pred, labels=['FAKE', 'REAL'])

array([[590,  48],
       [ 41, 588]])

- We have 589 true positive and 587 true negative. 
- We also have 42 false positives and 49 false negatives.

## Summary
Implementing a TfidVectorizer and implementing PassiveAggressiveClassifier, and fit our model. We ended up with a 92.98% accuracy. 